### ISSUE TITLE
The ActorReconstruction.testManyLocalSchedulersDying test in actor_test.py sometimes hangs.

### ISSUE THREAD

**Issue by robertnishihara** [ray-project/ray#1103]
If you run the following in a loop (on one of our Jenkins workers), one of the runs will eventually hang. This happens in Travis a lot as well.

```
python test/actor_test.py ActorReconstruction.testManyLocalSchedulersDying
```

Connecting another driver to one of the hanging instances, I see the following. Running

```python
import ray
ray.init(redis_address="127.0.0.1:27196")
tt = ray.global_state.task_table()
states = [v['State'] for k, v in tt.items()]
[(s, states.count(s)) for s in set(states)]
```

the last line prints

```
[(16, 694), (32, 143), (4, 20), (8, 2)]
```

so a number of tasks are in the `QUEUED` state. I checked and they seem to be queued at the live local scheduler (not a dead one). The number `4` corresponds to `QUEUED`. See 

https://github.com/ray-project/ray/blob/b1660c4edfcbfec5645103eaa0aad58a255015fa/src/common/task.h#L336-L352

I tried adding a commit to dispatch all tasks on a timer https://github.com/ray-project/ray/compare/master...robertnishihara:dispatchontimer, but that didn't change the behavior, tasks are still queued. I'm not totally sure what the problem is, but maybe the task dependencies or something are preventing some tasks from getting executed.

cc @stephanie-wang

**Comment by robertnishihara** [comment]
Ok, I've determined that the queued tasks are not getting dispatched because the call to `dispatch_actor_task` is returning in this if statement (note that this is a little behind the current master).

https://github.com/ray-project/ray/blob/b58500188143bbefb2aa358f32c9042885ffec8e/src/local_scheduler/local_scheduler_algorithm.cc#L315-L320

**Comment by stephanie-wang** [comment]
It'd be helpful to see the following information, if you can get it:
1. where the tasks with status 8 are scheduled.
2. whether the tasks with status 8 ever finish.
3. the values of `next_task_counter` and `entry.task_counter` at line 319.

**Comment by robertnishihara** [comment]
The tasks with status 8 are just dummy tasks for drivers.

In the case I am debugging right now, `entry.task_counter` is 0 (meaning we haven't executed any actor tasks yet), and `next_task_counter` is 1 (meaning that the first queued task is task 1). I printed out the queued tasks, and they have actor counters 1, 2, 3, ..., 80, but we are waiting for task 0.

Note that I have also seen situations where the queued tasks are more like 20, 21, 22, ..., 80, but we are waiting for task 0.

Task 0 is in state 32 (reconstructing). I've checked, and it looks like we are never calling `RAY.TASK_TABLE_TEST_AND_UPDATE` on task 0. We are repeatedly calling `RAY.TASK_TABLE_TEST_AND_UPDATE` on task 1, but it is getting suppressed because task 1 is QUEUED.

**Comment by robertnishihara** [comment]
So part of the issue is that we currently aren't calling `fetch_missing_dependencies` on actor tasks (which should be fixed by #1118).

However, I'm still a little surprised that we aren't attempting to reconstruct it when we recursively go through and reconstruct all of the missing dependencies for the object we are getting. Reconstruction must be getting suppressed. I could see this happening if the monitor has marked some actor tasks lost and simply hasn't gotten to the others yet, although that seems slightly uncommon but maybe not.

**Comment by robertnishihara** [comment]
Ok, we do attempt to reconstruct task 0 once, but it is suppressed because it's task status at that point is 2 (TASK_STATUS_SCHEDULED).

**My current favorite explanation is the following:**

Basically, reconstruction for all of the actor methods is initially suppressed because they are all in state SCHEDULED. Once the relevant local scheduler is killed, the monitor will slowly crawl over the task table and switch the statuses to LOST.

Now, because the local scheduler currently does not add the object dependencies of actor methods to the list of objects to fetch, normally reconstruction would not be reinitiated for these methods. However, the test is calling `ray.get` on all of the objects, so that is enough to reinitiate reconstruction for all the methods once the monitor has marked them as LOST.

HOWEVER, task 0 is the constructor task, which doesn't return an object ID to the user, so we are not calling `ray.get` on it, which means that we never reinitiate reconstruction for it.

Therefore, I think this problem should be completely solved by #1118.

**Comment by robertnishihara** [comment]
Closing because this should have been fixed by #1118.
